broadinstitute/scp-ingest-pipeline
# Author DE: Fix pairwise columns, support TSV (SCP-5292)

This fixes a scientific bug in transforming differential expression data, and improves robustness in a few areas.

Previously, columns would become misordered when the author DE file contained multiple pairwise comparisons that started with the same group.  For example, if the author DE file had pairwise comparisons for, say, "B cells and macrophages" _and_ "B cells and T cells", then the processed output for "B cells and macrophages" could contain metrics for "B cells and T cells".  Now, output is reliably ordered, and test data has been updated accordingly.

This also includes some scaffolding code for more flexible headers.  I found the bug while developing that, so -- as you can see in `de_utils.py` -- most of that code is commented out.  The current code retains the "basic" validation to ensure the author DE table explore UI doesn't break.

Finally, TSV data is now accepted as input.

To manually test:

1.  Run `python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 ingest_differential_expression --annotation-name General_Celltype --annotation-type group --annotation-scope study --cluster-name cluster_umap_txt --study-accession SCPdev --ingest-differential-expression --differential-expression-file gs://fc-febd4c65-881d-497f-b101-01a7ec427e6a/author_de_test/lfc_qval_scanpy-like.csv --method wilcoxon`
2. Run `ls -latr cluster_umap_txt--General_Celltype--B_cells--CSN1S1_macrophages--study--wilcoxon.tsv`
3. Confirm you see that file listed, with a timestamp from a few moments ago
4. Run `head -n 2 cluster_umap_txt--General_Celltype--B_cells--CSN1S1_macrophages--study--wilcoxon.tsv`
5. Confirm you see expected output of author DE pipeline
```
	genes	logfoldchanges	qval	mean
0	ACE2	0.685269917070053	0.446924681159184	0.727719408271226
```
6.  Confirm it matches the corresponding row from the raw, untransformed author DE file: run `grep "ACE2,B cells,CSN1S1 macrophages" /tmp/author_de_test%2flfc_qval_scanpy-like.csv`
7. Confirm you see `ACE2,B cells,CSN1S1 macrophages,0.6852699170700530,0.44692468115918400,0.7277194082712260`.

This satisfies SCP-5292.
